### DISCUSSION TITLE
Where should Service Objects go?

### DISCUSSION THREAD

**Discussion by nanobeep** [martenframework/marten#209] [Q&A] [answered]
First, really impressed with Marten, especially the documentation which is really excellent, thanks for all the effort and care put into it!

One question I couldn't find an answer to in the docs is how to handle Service Objects? In Rails and Django, there are conventions around them, but what would you suggest for Marten? Thanks in advance

(note: I also asked this in Discord and I'll update this discussion if the answer is given there, and vice-versa)

**Comment by ellmetha** [comment, marked as answer]
Hey! 👋 

Thanks for the kind words! 😊 

This is an excellent question. Presently there are no real conventions around this, and Marten projects can choose to to define service objects however they want! I am not opinionated regarding this aspect personally; if I had to define service objects in a Marten project, I would probably create a `services` folder under each application and define service abstraction under those.

An example for a project containing the default application only:

```
myproject/
├── config/
├── spec/
├── src
│   ├── assets/
│   ├── emails/
│   ├── handlers/
│   ├── migrations/
│   ├── models/
│   ├── schemas/
│   ├── services/
│   │   ├── object_creator.cr
│   │   └── event_notifier.cr
│   ├── templates/
│   ├── cli.cr
│   ├── project.cr
│   └── server.cr
├── manage.cr
└── shard.yml
```

And if the project contains multiple apps:

```
myproject/
├── config/
├── spec/
├── src
│   ├── apps/
│   │   ├── blog/
│   │   │   ├── handlers/
│   │   │   ├── migrations/
│   │   │   ├── models/
│   │   │   ├── schemas/
│   │   │   ├── services/
│   │   │   │   ├── article_creator.cr
│   │   │   │   └── tag_finder.cr
│   │   │   ├── templates/
│   │   │   ├── app.cr
│   │   │   ├── cli.cr
│   │   │   └── routes.cr
│   │   ├── users/
│   ├── cli.cr
│   ├── project.cr
│   └── server.cr
├── manage.cr
└── shard.yml
```

This is just an example, ultimately there are no guidelines or conventions around this aspect at the moment!
